adaptive/runner.py

```diff
@@ -861,3 +861,89 @@ def _get_ncores(ex):
         return ex._pool.size  # not public API!
     else:
         raise TypeError(f"Cannot get number of cores for {ex.__class__}")
+
+
+class _TimeGoal:
+    def __init__(self, dt: timedelta | datetime):
+        self.dt = dt
+        self.start_time = None
+
+    def __call__(self, _):
+        if isinstance(self.dt, timedelta):
+            if self.start_time is None:
+                self.start_time = datetime.now()
+            return datetime.now() - self.start_time > self.dt
+        elif isinstance(self.dt, datetime):
+            return datetime.now() > self.dt
+        else:
+            raise TypeError(f"{self.dt=} is not a datetime or timedelta.")
+
+
+def auto_goal(
+    goal: Callable[[BaseLearner], bool] | int | float | datetime | timedelta | None,
+    learner: BaseLearner,
+    allow_running_forever: bool = True,
+):
+    """Extract a goal from the learners.
+
+    Parameters
+    ----------
+    goal
+        The goal to extract. Can be a callable, an integer, a float, a datetime,
+        a timedelta or None.
+        If it is a callable, it is returned as is.
+        If it is an integer, the goal is reached after that many points have been
+        returned.
+        If it is a float, the goal is reached when the learner has reached a loss
+        less than that.
```

Comment:
This part seems fragile: the distinction between a float an an int isn't big enough.

Reply:
I just encountered this today https://scikit-learn.org/stable/modules/generated/sklearn.cluster.OPTICS.html#sklearn.cluster.OPTICS

<img width="943" alt="image" src="https://user-images.githubusercontent.com/6897215/197046577-cfbb60e5-2455-44b2-915c-f8eaaab1cf66.png">

Reply:
Fair point, however in this example also the ranges do not overlap. Loss, on the other hand, can assume arbitrary values.